Userland/Libraries/LibSoftGPU/SIMD.h

```diff
@@ -67,4 +67,60 @@ ALWAYS_INLINE static constexpr Vector4<AK::SIMD::i32x4> expand4(Vector4<int> con
     };
 }
 
+ALWAYS_INLINE static AK::SIMD::f32x4 ddx(AK::SIMD::f32x4 v)
+{
+    return AK::SIMD::f32x4 {
+        v[1] - v[0],
+        v[1] - v[0],
+        v[3] - v[2],
+        v[3] - v[2],
+    };
+}
+
+ALWAYS_INLINE static AK::SIMD::f32x4 ddy(AK::SIMD::f32x4 v)
+{
+    return AK::SIMD::f32x4 {
+        v[2] - v[0],
+        v[3] - v[1],
+        v[2] - v[0],
+        v[3] - v[1],
+    };
+}
+
+ALWAYS_INLINE static Vector2<AK::SIMD::f32x4> ddx(Vector2<AK::SIMD::f32x4> const& v)
+{
+    return {
+        ddx(v.x()),
+        ddx(v.y()),
+    };
+}
+
+ALWAYS_INLINE static Vector2<AK::SIMD::f32x4> ddy(Vector2<AK::SIMD::f32x4> const& v)
+{
+    return {
+        ddy(v.x()),
+        ddy(v.y()),
+    };
+}
+
+// approximation of log2 (https://stackoverflow.com/questions/9411823/fast-log2float-x-implementation-c)
+ALWAYS_INLINE static AK::SIMD::f32x4 log2_approximate(AK::SIMD::f32x4 v)
+{
+    union {
+        AK::SIMD::f32x4 float_val;
+        AK::SIMD::i32x4 int_val;
+    } u { v };
+
+    // extract just the exponent, giving lower integer bound for log2
+    auto log = AK::SIMD::to_f32x4(((u.int_val >> 23) & 255) - 128);
+
+    // replace the exponent with 127 (0)
+    u.int_val &= ~(255 << 23);
+    u.int_val += 127 << 23;
```

Comment:
In the context of bitwise operations, I'd expect this to be an `|=` instead of `+=` (but both work since it does not carry any bits during the addition).